src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActions/SuggestedAction.cs

```diff
@@ -96,82 +98,103 @@ protected Task<ImmutableArray<CodeActionOperation>> GetPreviewOperationsAsync(Ca
-                    actionCallback.Value.OnSuggestedActionExecuted(this);
+                using var scope = context.AddScope(allowCancellation: true, CodeAction.Message);
+                await this.InvokeAsync(new UIThreadOperationContextProgressTracker(scope), context.UserCancellationToken).ConfigureAwait(false);
             }
+            catch (OperationCanceledException)
+            {
+            }
+            catch (Exception ex) when (FatalError.ReportAndCatch(ex))
+            {
+            }
+        }
+
+        private async Task InvokeAsync(IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            await InnerInvokeAsync(progressTracker, cancellationToken).ConfigureAwait(true);
+            foreach (var actionCallback in SourceProvider.ActionCallbacks)
+                actionCallback.Value.OnSuggestedActionExecuted(this);
         }
 
-        protected virtual void InnerInvoke(IProgressTracker progressTracker, CancellationToken cancellationToken)
+        protected virtual async Task InnerInvokeAsync(IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
+            // Even though we're async, we should still be on the UI thread at this point.
             AssertIsForeground();
 
             var snapshot = SubjectBuffer.CurrentSnapshot;
             using (new CaretPositionRestorer(SubjectBuffer, EditHandler.AssociatedViewService))
             {
                 Document getFromDocument() => SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-                InvokeCore(getFromDocument, progressTracker, cancellationToken);
+                await InvokeCoreAsync(getFromDocument, progressTracker, cancellationToken).ConfigureAwait(false);
             }
         }
 
-        protected void InvokeCore(
+        protected async Task InvokeCoreAsync(
             Func<Document> getFromDocument, IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
+            // Even though we're async, we should still be on the UI thread at this point.
             AssertIsForeground();
 
             var extensionManager = Workspace.Services.GetService<IExtensionManager>();
-            extensionManager.PerformAction(Provider, () =>
+            await extensionManager.PerformActionAsync(Provider, async () =>
             {
-                InvokeWorker(getFromDocument, progressTracker, cancellationToken);
-            });
+                await InvokeWorkerAsync(getFromDocument, progressTracker, cancellationToken).ConfigureAwait(false);
+            }).ConfigureAwait(false);
         }
 
-        private void InvokeWorker(
+        private async Task InvokeWorkerAsync(
             Func<Document> getFromDocument, IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
+            // Even though we're async, we should still be on the UI thread at this point.
```

Comment:
💡 This is not a customary pattern. All of these should eventually transition to calls to `SwitchToMainThreadAsync`

Reply:
true.  i wanted this though to at least know if my expectations were being broken.  but i will switch these.